include/session/network.hpp

```diff
@@ -0,0 +1,217 @@
+#pragma once
+
+#include <oxen/log/ring_buffer_sink.hpp>
+#include <oxen/quic.hpp>
+
+#include "network_service_node.hpp"
+#include "session/onionreq/builder.hpp"
+#include "session/onionreq/key_types.hpp"
+#include "session/types.hpp"
+
+namespace session::network {
+
+enum class ServiceNodeChangeType {
+    none = 0,
+    invalid_path = 1,
+    replace_swarm = 2,
+    update_path = 3,
+    update_node = 4,
+};
+
+struct onion_path {
+    std::shared_ptr<oxen::quic::connection_interface> conn;
+    std::vector<session::network::service_node> nodes;
+    uint8_t failure_count;
+};
+
+struct service_node_changes {
+    ServiceNodeChangeType type = ServiceNodeChangeType::none;
+    std::vector<session::network::service_node> nodes = {};
+    uint8_t path_failure_count = 0;
+    bool path_invalid = false;
+};
+
+struct request_info {
+    const service_node target;
+    const std::string endpoint;
+    const std::optional<ustring> body;
+    const std::optional<std::vector<service_node>> swarm;
+    const std::optional<onion_path> path;
+    const bool is_retry;
```

Comment:
> But I think this won't be sufficient because of other changes along the way, and so I think we need to rework this a little.

To elaborate: I left a comment here about not using a promise/future, and I think to make that work we'll need to make a copy of this object.

Reply:
I've removed the `const` from these - now that the snode pool is internal it might be easier to see how this is used (it's essentially only used to allow for the `retry` request in the case of receiving a `421`)